repo2docker/buildpacks/r.py

```diff
@@ -0,0 +1,211 @@
+import re
+import os
+import datetime
+
+from .python import PythonBuildPack
+
+class RBuildPack(PythonBuildPack):
+    """
+    Setup R for use with a repository
+
+    This sets up R + RStudio + IRKernel for a repository that contains:
+
+    1. A `runtime.txt` file with the text:
+
+       r-<year>-<month>-<date>
+
+       Where 'year', 'month' and 'date' refer to a specific
+       date snapshot of https://mran.microsoft.com/timemachine
+       from which libraries are to be installed.
+
+    2. An optional `install.R` file that will be executed at build time,
+       and can be used for installing packages from both MRAN and GitHub.
+
+    The `r-base` package from Ubuntu apt repositories is used to install
+    R itself, rather than any of the methods from https://cran.r-project.org/.
+    """
+    @property
+    def runtime(self):
+        """
+        Return contents of runtime.txt if it exists, '' otherwise
+        """
+        if not hasattr(self, '_runtime'):
+            runtime_path = self.binder_path('runtime.txt')
+            try:
+                with open(runtime_path) as f:
+                    self._runtime = f.read().strip()
+            except FileNotFoundError:
+                self._runtime = ''
+
+        return self._runtime
+
+    @property
+    def checkpoint_date(self):
+        """
+        Return the date of MRAN checkpoint to use for this repo
+
+        Returns '' if no date is specified
+        """
+        if not hasattr(self, '_checkpoint_date'):
```

Comment:
If you wanted to get rid of the repetitive `hasattr()` tests in all the properties you could use [`@lru_cache`](https://docs.python.org/3/library/functools.html#functools.lru_cache):
```python
from functools import lru_cache as cache

@property
@cache(1)
def checkpoint_date(self):
    match = ...
```

There might be a neat way to turn `@cache(1)` into just `@cache` via `cache = functools.partial(lru_cache, 1)`, but maybe that is too much magic :-/

Reply:
IMO using @cache etc is pretty opaque compared with just using `hasattr`

I just pushed a PR to add a short README to the R test, which should make it show up in the `supported configuration files` section when the docs are built.

I'm +1 on merge - I think we should get this into the wild so we can start spot-checking improvements